collect_app/src/main/java/org/odk/collect/android/formmanagement/ServerFormDownloader.java

```diff
@@ -58,14 +58,16 @@ public ServerFormDownloader(FormSource formSource, FormsRepository formsReposito
 
     @Override
     public void downloadForm(ServerFormDetails form, @Nullable ProgressReporter progressReporter, @Nullable Supplier<Boolean> isCancelled) throws FormDownloadException, InterruptedException {
-        Form formOnDevice = formsRepository.getOneByFormIdAndVersion(form.getFormId(), form.getFormVersion());
+        Form formOnDevice = formsRepository.getOneByMd5Hash(getMd5HashWithoutPrefix(form.getHash()));
         if (formOnDevice != null) {
             if (formOnDevice.isDeleted()) {
                 formsRepository.restore(formOnDevice.getId());
             } else {
-                String formIdentifier = formOnDevice.getDisplayName() + " " + formOnDevice.getId();
-                String formIdHash = FileUtils.getMd5Hash(new ByteArrayInputStream(formIdentifier.getBytes()));
-                analytics.logFormEvent(DOWNLOAD_SAME_FORMID_VERSION, formIdHash);
+                if (!getMd5HashWithoutPrefix(form.getHash()).equals(formOnDevice.getMD5Hash()) && !form.getDownloadUrl().contains("/draft.xml")) {
```

Comment:
I could be confused, but I don't think this will ever be `true` right? Here we're looking up `formOnDevice` using `form`'s hash and then in this else you'd be comparing them (so they'd always match and the `if` would be false). Maybe it's worth having a test for this analytics call as it's a bit fiddly.

Sorry if I'm totally wrong about this and just need to wake up a little more!

Reply:
Sigh. You're absolutely right. That was me not coming back to what I was actually trying to accomplish after addressing your feedback. It used to make sense before the condition for undeleting changed.